drivers/sensor/bosch/bmm350/bmm350_bus.c

```diff
@@ -0,0 +1,136 @@
+/*
+ * Copyright (c) 2024 Bosch Sensortec GmbH
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "bmm350.h"
+
+static int bmm350_bus_check_rtio(const struct bmm350_bus *bus)
+{
+	if (bus->rtio.type == BMM350_BUS_TYPE_I2C) {
+		return i2c_is_ready_iodev(bus->rtio.iodev) ? 0 : -ENODEV;
+	}
+
+	return -EIO;
+}
+
+static int bmm350_prep_reg_read_rtio_async(const struct bmm350_bus *bus,
+					   uint8_t reg, uint8_t *buf, size_t size,
+					   struct rtio_sqe **out)
+{
+	struct rtio *ctx = bus->rtio.ctx;
+	struct rtio_iodev *iodev = bus->rtio.iodev;
+	struct rtio_sqe *write_reg_sqe = rtio_sqe_acquire(ctx);
+	struct rtio_sqe *read_buf_sqe = rtio_sqe_acquire(ctx);
+
+	if (!write_reg_sqe || !read_buf_sqe) {
+		rtio_sqe_drop_all(ctx);
+		return -ENOMEM;
+	}
+
+	rtio_sqe_prep_tiny_write(write_reg_sqe, iodev, RTIO_PRIO_NORM, &reg, 1, NULL);
+	write_reg_sqe->flags |= RTIO_SQE_TRANSACTION;
+	rtio_sqe_prep_read(read_buf_sqe, iodev, RTIO_PRIO_NORM, buf, size, NULL);
+	if (bus->rtio.type == BMM350_BUS_TYPE_I2C) {
+		read_buf_sqe->iodev_flags |= RTIO_IODEV_I2C_STOP | RTIO_IODEV_I2C_RESTART;
+	}
+
+	/** Send back last SQE so it can be concatenated later. */
+	if (out) {
+		*out = read_buf_sqe;
+	}
+
+	return 2;
+}
+
+static int bmm350_prep_reg_write_rtio_async(const struct bmm350_bus *bus,
+					    uint8_t reg, uint8_t val,
+					    struct rtio_sqe **out)
+{
+	struct rtio *ctx = bus->rtio.ctx;
+	struct rtio_iodev *iodev = bus->rtio.iodev;
+	struct rtio_sqe *write_reg_sqe = rtio_sqe_acquire(ctx);
+	struct rtio_sqe *write_buf_sqe = rtio_sqe_acquire(ctx);
+
+	if (!write_reg_sqe || !write_buf_sqe) {
+		rtio_sqe_drop_all(ctx);
+		return -ENOMEM;
+	}
+
+	rtio_sqe_prep_tiny_write(write_reg_sqe, iodev, RTIO_PRIO_NORM, &reg, 1, NULL);
+	write_reg_sqe->flags |= RTIO_SQE_TRANSACTION;
+	rtio_sqe_prep_tiny_write(write_buf_sqe, iodev, RTIO_PRIO_NORM, &val, 1, NULL);
+	if (bus->rtio.type == BMM350_BUS_TYPE_I2C) {
+		write_buf_sqe->iodev_flags |= RTIO_IODEV_I2C_STOP;
+	}
+
+	/** Send back last SQE so it can be concatenated later. */
+	if (out) {
+		*out = write_buf_sqe;
+	}
+
+	return 2;
+}
+
+static int bmm350_reg_read_rtio(const struct bmm350_bus *bus, uint8_t start, uint8_t *buf, int size)
+{
+	struct rtio *ctx = bus->rtio.ctx;
+	struct rtio_cqe *cqe;
+	int err;
+
+	err = bmm350_prep_reg_read_rtio_async(bus, start, buf, size, NULL);
+	if (err < 0) {
+		return err;
+	}
+
+	err = rtio_submit(ctx, err);
+	if (err) {
+		return err;
+	}
+
+	do {
+		cqe = rtio_cqe_consume(ctx);
+		if (cqe != NULL) {
+			err = cqe->result;
+			rtio_cqe_release(ctx, cqe);
+		}
+	} while (cqe != NULL);
+
+	return err;
+}
+
+static int bmm350_reg_write_rtio(const struct bmm350_bus *bus, uint8_t reg, uint8_t val)
+{
+	struct rtio *ctx = bus->rtio.ctx;
+	struct rtio_cqe *cqe;
+	int err;
+
+	err = bmm350_prep_reg_write_rtio_async(bus, reg, val, NULL);
+	if (err < 0) {
+		return err;
+	}
+
+	err = rtio_submit(ctx, err);
```

Comment:
re-use of the err var name here looks a bit funny, ret might be more reasonable for the var name without the suspicious looking rtio_submit(ctx, err)